File: Programmers/pro_1844_게임 맵 최단거리.py
from collections import deque
def solution(maps):
    R = len(maps)
    C = len(maps[0])
    visited=[[0 for _ in range(C)] for _ in range(R)]
    
    move1 = [1,0,0,-1]
    move2 = [0,1,-1,0]
    
    que = deque([(0,0)])
    visited[0][0]=1
    answer = 0
    while(que):
        answer+=1
        for i in range(len(que)):
            topR, topC = que.popleft()
            for dr, dc in zip(move1,move2):
                nextR = topR + dr
                nextC = topC + dc
                if(0<=nextR<R and 0<=nextC<C and maps[nextR][nextC]==1 and visited[nextR][nextC]==0):
                    if(nextR==R-1 and nextC==C-1):
                        return answer + 1
                    visited[nextR][nextC]=1
                    que.append((nextR,nextC))
    return -1

File: Programmers/pro_1844_게임맵최단거리.js
function solution(maps) {
    const [R, C] = [maps.length, maps[0].length]
    const visited = Array.from(Array(R), () => Array.from(Array(C), () => 0))
    const que = [[0, 0]]
    const move = [[1, 0], [0, 1], [-1, 0], [0, -1]]

    let answer = 1;
    let loop = 0
    while (que.length) {
        loop += 1
        let len = que.length
        while (len--) {
            let [top_r, top_c] = que.shift()
            for (let i = 0; i < 4; i++) {
                let [dr, dc] = move[i]
                let [nextR, nextC] = [top_r + dr, top_c + dc]
                if (nextR < 0 || R <= nextR || nextC < 0 || C <= nextC)
                    continue
                if (maps[nextR][nextC] == 1 && visited[nextR][nextC] == 0) {
                    if (nextR == R - 1 && nextC == C - 1)
                        return answer + 1
                    visited[nextR][nextC] = 1
                    que.push([nextR, nextC])
                }
            }
        }
        answer += 1
    }
    return -1;
}